main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -134,24 +134,23 @@ object EntryPoint {
       // First check that there's exactly one argument.
       val argVal = declaredScheme.base.arrowArgTypes match {
         // Case 1: One arg. Ok :)
-        case arg :: Nil => arg.toSuccess
+        case arg :: Nil => Some(arg).toSuccess
         // Case 2: Multiple args. Error.
-        case _ :: _ :: _ => EntryPointError.IllegalEntryPointArgs(sym, sym.loc).toFailure
+        case _ :: _ :: _ => None.toSoftFailure(EntryPointError.IllegalEntryPointArgs(sym, sym.loc))
         // Case 3: Empty arguments. Impossible since this is desugared to Unit.
         case Nil => throw InternalCompilerException("Unexpected empty argument list.", loc)
       }
 
-      flatMapN(argVal: Validation[Type, EntryPointError]) {
-        arg =>
-          val argSc = Scheme.generalize(Nil, Nil, arg)
-
-          if (Scheme.equal(unitSc, argSc, classEnv, ListMap.empty)) { // TODO ASSOC-TYPES better eqEnv
-            // Case 1: Unit -> XYZ. We can ignore the args.
-            ().toSuccess
-          } else {
-            // Case 2: Bad arguments. Error.
-            EntryPointError.IllegalEntryPointArgs(sym, sym.loc).toFailure
-          }
+      flatMapN(argVal: Validation[Option[Type], EntryPointError]) {
+        // Case 1: Unit -> XYZ. We can ignore the args.
+        case Some(arg) if Scheme.equal(unitSc, Scheme.generalize(Nil, Nil, arg), classEnv, ListMap.empty) =>
+          // TODO ASSOC-TYPES better eqEnv
+          ().toSuccess
+
+        // Case 2: Bad arguments. Error
+        // Case 3: argVal was None / SoftError
+        case Some(_) | None =>
```

Comment:
This code is really ugly. Can we not do better?

Reply:
:)

Reply:
Oh oh... what style do you prefer / is better?

Reply:
I just find the whole pattern match so confusing. Like when does whole succeed or fail?

Reply:
@mlutze

Reply:
Better with a wildcard?